src/test/java/dog/pawbook/logic/parser/EditProgramCommandParserTest.java

```diff
@@ -0,0 +1,166 @@
+
+        // multiple invalid values, but only the first invalid value is captured
+        assertParseFailure(parser, "1" + INVALID_NAME_DESC + INVALID_SESSION_DESC + VALID_TAG_ALL,
+                Name.MESSAGE_CONSTRAINTS);
+    }
+
+    @Test
+    public void parse_allFieldsSpecified_success() {
+        Integer targetId = ID_TWO;
+        String userInput = targetId + TAG_DESC_ALL + SESSION_DESC_OBEDIENCE_TRAINING
+                + NAME_DESC_OBEDIENCE_TRAINING + TAG_DESC_PUPPIES;
+
+        EditProgramDescriptor descriptor = new EditProgramDescriptorBuilder().withName(VALID_NAME_OBEDIENCE_TRAINING)
+                .withSessions(VALID_SESSION_OBEDIENCE_TRAINING).withTags(VALID_TAG_ALL, VALID_TAG_PUPPIES).build();
+        EditProgramCommand expectedCommand = new EditProgramCommand(targetId, descriptor);
+
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_someFieldsSpecified_success() {
+        Integer targetId = ID_ONE;
+        String userInput = targetId + SESSION_DESC_OBEDIENCE_TRAINING + NAME_DESC_POTTY_TRAINING;
+
+        EditProgramDescriptor descriptor = new EditProgramDescriptorBuilder().withName(VALID_NAME_POTTY_TRAINING)
+                .withSessions(VALID_SESSION_OBEDIENCE_TRAINING).build();
+        EditProgramCommand expectedCommand = new EditProgramCommand(targetId, descriptor);
+
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_oneFieldSpecified_success() {
+        // name
+        Integer targetId = ID_THREE;
+        String userInput = targetId + NAME_DESC_OBEDIENCE_TRAINING;
+        EditProgramDescriptor descriptor = new EditProgramDescriptorBuilder()
+                .withName(VALID_NAME_OBEDIENCE_TRAINING).build();
+        EditProgramCommand expectedCommand = new EditProgramCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+
+        // sessions
+        userInput = targetId + SESSION_DESC_OBEDIENCE_TRAINING;
+        descriptor = new EditProgramDescriptorBuilder().withSessions(VALID_SESSION_OBEDIENCE_TRAINING).build();
+        expectedCommand = new EditProgramCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+
+        // tags
+        userInput = targetId + TAG_DESC_ALL;
+        descriptor = new EditProgramDescriptorBuilder().withTags(VALID_TAG_ALL).build();
+        expectedCommand = new EditProgramCommand(targetId, descriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_multipleRepeatedFields_acceptsLast() {
+        Integer targetId = ID_ONE;
+        String userInput = targetId + TAG_DESC_PUPPIES + TAG_DESC_ALL + TAG_DESC_ALL + TAG_DESC_PUPPIES
+                + SESSION_DESC_POTTY_TRAINING + SESSION_DESC_POTTY_TRAINING + TAG_DESC_PUPPIES;
```

Comment:
I think we need another test that ensures that those repeatable fields like tags etc, if different values are supplied, all will be captured, and if repeated, it will be deduplicated correctly.

Reply:
Wait, I thought this clearly covers the duplicated tags and different tags? I can add different sessions if that will help.

Reply:
I'm referring to more test cases that can be written, this line is the most logical thing I can point to. For tags and sessions, repeated values are all captured, but the tests clearly don't look out for this, in this current test case there is only repeated tags, which will be parsed as one.

Reply:
Done!